test/dataset/test_sampling.py

```diff
@@ -41,6 +41,25 @@ def test_single_cut_sampler_shuffling():
     assert sampler_cut_ids != [c.id for c in cut_set]
 
 
+def test_single_cut_sampler_len():
```

Comment:
can these tests check that the batch ordering is different for every epoch too? you can just store the "batches" as lists and assert that they are not equal (should work since these are just collections of string ids, not dicts with actual data)

Reply:
(this is mostly to avoid regressions in the future, I can see that the current code should work as expected)

Reply:
I think it would be good to also have a test that shows that the order is deterministic given the same epoch number. This is not an obvious property in general and I wouldn't be surprised if we accidentally break it trying to update this code in the future.

Reply:
I added several tests for ordering and determinism. The sampling of buckets in bucket sampler is now also deterministic and dependent on seed and epoch. I also added seed parameter for the bucket sampler to be able to control the bucket sampling itself.

One more idea: We could add a parameter that would automatically increase epoch after iteration so that user does not have to manually call `set_epoch`. WYT?

Reply:
Hm and one more problem - the `list(sampler)` works fine in `python3.8` but not in `python3.7` for some reason. I don't understand what exactly is wrong.

Reply:
> One more idea: We could add a parameter that would automatically increase epoch after iteration so that user does not have to manually call `set_epoch`. WYT?

I'd rather not - that would be inconsistent with what PyTorch's DistributedSampler does.

> Hm and one more problem - the list(sampler) works fine in python3.8 but not in python3.7 for some reason. I don't understand what exactly is wrong.

I wouldn't worry about that - if this is problematic later we can look into it.